designs/extensible-cli-and-scaffolding-plugins-phase-2.md

```diff
@@ -0,0 +1,326 @@
+
+* Pros
+  * This provides flexibility for the user to specify the file paths that the plugin would be placed in and `kubebuilder` could discover the binaries in those user specified file paths.
+
+  * No constraints on plugin binary naming or directory placements from the Kubebuilder side.
+
+  * Provides a default value for the plugin directory in case user wants to use that to drop their plugins.
+
+#### Prefixed plugin executable names in $PATH
+
+Another approach is adding plugin executables with a prefix `kubebuilder-` followed by the plugin name to the PATH variable. This will enable `kubebuilder` to traverse through the PATH looking for the plugin executables starting with the prefix `kubebuilder-` and matching by the plugin name that was provided in the CLI. Furthermore, a check should be added to verify that the match is an executable or not and return an error if it's not an executable. This approach provides a lot of flexibility in terms of plugin discovery as all the user needs to do is to add the plugin executable to the PATH and `kubebuilder` will discover it.
+
+* Pros
+  * `kubectl` and `git` follow the same approach for discovering plugins, so there’s prior art.
+
+  * There’s a lot of flexibility in just dropping plugin binaries to PATH variable and enabling the discovery without having to enforce any other constraints on the placements of the plugins.
+
+* Cons
+  * Enumerating the list of all available plugins might be a bit tough compared to having a single folder with the list of available plugins and having to enumerate those.
+
+  * These plugin binaries cannot be run in a standalone manner outside of Kubebuilder, so may not be very ideal to add them to the PATH var.
+
+## Open questions
+
+* Do we want to support the addition of new arbitrary subcommands other than the subcommands (init, create api, create webhook) that we already support?
+  * Not for the EP or initial implementation, but can revisit later.
+
+* Do we need to discover flags by calling the plugin binary or should we have users define them in the project configuration?
+  * Flags will be passed directly to the external plugin as a string. Flag parse errors will be passed back via `PluginResponse`.
+
+* What alternatives to stdin/stdout exist and why shouldn't we use them?
+  * Other alternatives exist such as named pipe and sockets, but stdin/stdout seems to be more suitable for our needs.
+
+* What happens when two plugins bind the same flag name? Will there be any conflicts?
+  * Flags are passed directly as a string to plugins. The same string will be passed to each plugin in the chain, so all plugins get the same flag set. Errors should not be returned if an unrecognized flag is parsed.
```

Comment:
Are we documenting that to plugin writers?

Reply:
I mentioned this in the Implementation Details. Is there some other place to add this that would be more meaningful?

Reply:
I just missed it. Thanks.